gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/src/AuthorizationService.js

```diff
@@ -662,42 +712,125 @@ var fluid = fluid || require("infusion");
-                });
-                fluid.promise.follow(addAuthDecisionPromise, promiseTogo);
-            }
-        }, function (err) {
-            promiseTogo.reject(err);
+
+        var tokenPromise = dataStore.findResourceOwnerTokenByGpiiTokenAndClientId(gpiiToken, clientId);
+
+        tokenPromise.then(function (tokens) {
+            var setExpiredPromises = [];
+            var tokenObj;
+
+            fluid.each(tokens, function (token) {
+                if (gpii.oauth2.isExpired(token.timestampCreated, token.expiresIn)) {
+                    setExpiredPromises.push(dataStore.expireResourceOwnerToken(token.id));
+                } else {
+                    tokenObj = token;
+                }
+            });
+            var setExpiredSequence = fluid.promise.sequence(setExpiredPromises);
+
+            setExpiredSequence.then(function () {
+                var result = tokenObj ? {
+                    accessToken: tokenObj.accessToken,
+                    expiresIn: tokenObj.expiresIn
+                } : undefined;
+
+                promiseTogo.resolve(result);
+            });
+        });
+
+        return promiseTogo;
+    };
+
+    /**
+     * @param promiseTogo {Object} Modified by the function with objects to be resolved or to fail
+     * @param dataStore {Component} An instance of gpii.oauth2.dbDataStore
+     * @param codeGenerator {Component} An instance of gpii.oauth2.codeGenerator
+     * @param gpiiToken {String} a GPII token
+     * @param clientId {String} an unique client id
+     * @param expiresIn {String} the number of seconds that this token will expire
+     * @return: none. The first argument of promiseTogo contains returned values
+     */
+    gpii.oauth2.authorizationService.createResourceOwnerToken = function (promiseTogo, dataStore, codeGenerator, gpiiToken, clientId, expiresIn) {
+        var accessToken = codeGenerator.generateAccessToken();
+
+        var addResourceOwnerTokenPromise = dataStore.addResourceOwnerToken({
+            clientId: clientId,
+            gpiiToken: gpiiToken,
+            accessToken: accessToken,
+            expiresIn: expiresIn
         });
+
+        var mapper = function () {
+            // The created access token is resolved for promiseTogo eventually
+            return {
+                accessToken: accessToken,
+                expiresIn: expiresIn
+            };
+        };
+        var accessTokenPromise = fluid.promise.map(addResourceOwnerTokenPromise, mapper);
+        fluid.promise.follow(accessTokenPromise, promiseTogo);
+    };
+
+    /**
+     * Grant a resource owner password credential access token. The gpii token will be verified before the access token is returned.
+     * @param dataStore {Component} An instance of gpii.oauth2.dbDataStore
+     * @param codeGenerator {Component} An instance of gpii.oauth2.codeGenerator
+     * @param clientId {String} A client id
+     * @param gpiiToken {String} A GPII token
```

Comment:
Order here differs from function param list. `clientId` and `gpiiToken` are reversed.